proposals/csharp-12.0/collection-expressions.md

```diff
@@ -219,7 +213,79 @@ ImmutableArray<int> ia =
+    * An applicable `AddRange` instance or extension method is invoked on the *collection instance* with the spread element *expression* as the argument.
+    * An applicable `CopyTo` instance or extension method is invoked on the *spread element expression* with the collection instance and `int` index as arguments.
+
+* During the construction steps above, an applicable `EnsureCapacity` instance or extension method *may* be invoked one or more times on the *collection instance* with an `int` capacity argument.
+
+---
+
+If the target type is an *array*, a *span*, a type with a *[create method](#create-methods)*, or an *interface*, the construction of the collection instance is as follows:
+
+* The elements are evaluated in order. Some or all elements may be evaluated *during* the steps below rather than before.
+
+* The compiler *may* determine the *known length* of the collection expression by invoking [*countable*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/ranges.md#adding-index-and-range-support-to-existing-library-types) properties &mdash; or equivalent properties from well-known interfaces or types &mdash; on each *spread element expression*.
+
+* A collection instance is created as follows:
+  * If the target type is an *array* and the collection expression has a *known length*, an array is allocated with the expected length.
+  * If the target type is a *span* or a type with a *create method*, and the collection has a *known length*, a span with the expected length is created referring to allocated storage.
+  * Otherwise intermediate storage is allocated.
+
+* For each element in order:
+  * If the element is an *expression element*, the collection *indexer* is invoked to add the evaluated expression at the current index.
+  * If the element is a *spread element* then one of the following is used:
+    * A member of a well-known interface or type is invoked to copy items from the spread element expression to the collection instance.
+    * An applicable `GetEnumerator` instance or extension method is invoked on the *spread element expression* and for each item from the enumerator, the collection *indexer* is invoked to add the item at the current index. If the enumerator implements `IDisposable`, then `Dispose` will be called after enumeration, regardless of exceptions.
+    * An applicable `CopyTo` instance or extension method is invoked on the *spread element expression* with the collection instance and `int` index as arguments.
+
+* If intermediate storage was allocated for the collection, a collection instance is allocated with the actual length, and the values from the intermediate storage are copied to the target collection instance.
```

Comment:
1. perhaps mention that if the length is correct, no copy is needed.
2. for array, i get why we need to copy, so that part is fine.  However, for span, it seems like we could just point the span at the portion of the array we filled.  Given that the span will be short lived (Since it can only be on the stack), that might be nice to do instead.

Reply:
The "intermediate storage" case is only for when the collection does not have a known length (see line 255).

Reply:
note: even for the case with a span destination, we likely still don't want a final copy of the array.  INstead, we should just make a span over the subset of the array containing elements.  

In other words, say we're using an array, starting at 8 elements large, and growing it by doubling.

If we add a total of 10 elements, we should double once (to 16), then get a span pointing to the first 10 elements.  We should *not* take those 16-element-array and copy out the 10 elements into a 10-element array that is then wrapped by a span.

Does that make sense?